umbrella-feeds/src/lib.rs

```diff
@@ -0,0 +1,732 @@
+
+        ensure!(
+            prev_signer < Some(signer),
+            CustomContractError::SignaturesOutOfOrder
+        );
+
+        validators.push(signer);
+
+        prev_signer = Some(signer);
+    }
+
+    let are_valid_signers = host.invoke_contract_read_only::<Vec<AccountAddress>>(
+        &host.state().staking_bank,
+        &validators,
+        EntrypointName::new_unchecked("verifyValidators"),
+        Amount::zero(),
+    )?;
+
+    let are_valid_signers: bool = are_valid_signers
+        .ok_or(CustomContractError::InvokeContractError)?
+        .get()?;
+
+    ensure_eq!(are_valid_signers, true, CustomContractError::InvalidSigner);
+
+    Ok(())
+}
+
+#[receive(
+    contract = "umbrella_feeds",
+    name = "update",
+    parameter = "UpdateParams",
+    error = "CustomContractError",
+    crypto_primitives,
+    mutable
+)]
+fn update<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    crypto_primitives: &impl HasCryptoPrimitives,
+) -> Result<(), CustomContractError> {
+    let param: UpdateParams = ctx.parameter_cursor().get()?;
+
+    let message = param.message;
+
+    // Check that the signature was intended for this contract.
+    ensure_eq!(
+        message.contract_address,
+        ctx.self_address(),
+        CustomContractError::WrongContract
+    );
+
+    // Check signature is not expired.
+    ensure!(
+        message.timestamp > ctx.metadata().slot_time(),
+        CustomContractError::Expired
+    );
+
+    // Check signature has correct chain_id.
+    ensure_eq!(
+        message.chain_id,
+        CHAIN_ID,
+        CustomContractError::ChainIdMismatch
+    );
+
+    verify_signatures(ctx, host, crypto_primitives)?;
+
+    for element in message.price_feed {
+        let price_key: HashSha2256 = element.0;
+        let new_price_data: PriceData = element.1;
+
+        let mut stored_price_data = host
+            .state_mut()
+            .prices
+            .entry(price_key)
+            .or_insert_with(PriceData::default);
+
+        // We do not allow for older prices.
+        // This prevents replay attacks by preventing reusing of signatures at the same time.
+        ensure!(
+            stored_price_data.timestamp < new_price_data.timestamp,
+            CustomContractError::OldData
+        );
+
+        *stored_price_data = new_price_data;
+    }
+
+    Ok(())
+}
+
+/// View function that returns the key hash/name of this contract.
+#[receive(
+    contract = "umbrella_feeds",
+    name = "getName",
+    return_value = "HashSha2256",
+    crypto_primitives
+)]
+fn get_name<S: HasStateApi>(
+    _ctx: &impl HasReceiveContext,
+    _host: &impl HasHost<State<S>, StateApiType = S>,
+    crypto_primitives: &impl HasCryptoPrimitives,
+) -> ReceiveResult<HashSha2256> {
+    let key_hash = crypto_primitives.hash_sha2_256(NAME.as_bytes()).0;
+
+    Ok(HashSha2256(key_hash))
```

Comment:
Since this returns a fixed hash always, is there a reason why it is computed each time?